### PR TITLE
Retrofit Integration.

### PR DESCRIPTION
**Pull request by johnrengelman**
This change is [Reviewable](https://reviewable.io/reviews/ratpack/ratpack/976)

### CODE REVIEW

**Comment by johnrengelman** [comment]
@ratpack/committers would appreciate some feedback, especially around the API in `RatpackRetrofit`

File: ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  public static class Builder {
+
+    private HttpClient httpClient;
+    private Action<? super Retrofit.Builder> builderAction = Action.noop();
+    private URI uri;
+
+    public Builder(HttpClient httpClient) {
+      this.httpClient = httpClient;
```
Comment (johnrengelman) [inline ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java]: I think maybe I can do away with this by using `Execution.current().get(HttpClient.class)` down when the request is actually made?

**Comment by alkemist** [comment]
Reviewed 2 of 13 files at r1, 1 of 1 files at r2.
Review status: 3 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[ratpack-core/src/main/java/ratpack/service/DependsOn.java, line 70 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3rndyLkQ9G_ZO9OD:-KIX3rndyLkQ9G_ZO9OE:1810560380) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/DependsOn.java#L70)):*
> ```Java
>  *   public static void main(String[] args) throws Exception {
>  *     RatpackServer server = RatpackServer.of(s -> s
>  *       .serverConfig(ServerConfig.embedded())
> ```

Why was this needed?

---

*[ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java, line 72 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3twZar7PpKmKebTo:-KIX3twZar7PpKmKebTp:1810560380) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java#L72)):*
> ```Java
>  *   public static void main(String[] args) throws Exception {
>  *     RatpackServer server = RatpackServer.of(s -> s
>  *       .serverConfig(ServerConfig.embedded())
> ```

Why was this needed?

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4WoZu3-LZ4BRKDV5:-KIX4WoZu3-LZ4BRKDV6:1959722109) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-manual/src/content/chapters/38-retrofit.md#L50)):*
> ```Markdown
>           PublicAddress address = ctx.get(PublicAddress.class);
>           
>           HelloApi api = RatpackRetrofit
> ```

I think we should be setting up these objects in the registry for these examples, as that's closer to reality and provides a better hint on what people should do.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java, line 96 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KGsOorVnHsVhWkXxoye-r1-96:-KIX4yCuHNyCQ1rxRcth:-1572895293) ([raw file](https://github.com/ratpack/ratpack/blob/9ebc95889df11d68d7e38cddc0e00e8249c0013d/ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java#L96)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> I think maybe I can do away with this by using `Execution.current().get(HttpClient.class)` down when the request is actually made?
</details>
Do away with what? asking for a URI?

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java, line 74 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4r_U4QcE-b2ibrbQ:-KIX4r_U4QcE-b2ibrbR:-875504425) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java#L74)):*
> ```Java
>  *
>  */
> public class RatpackRetrofit {
> ```

should be abstract.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java, line 140 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX5AdKNcoD2zGk1C_I:-KIX5AdKNcoD2zGk1C_J:-484069531) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java#L140)):*
> ```Java
> 
>     /**
>      * Uses this builder to create a Retrofit client implemention.
> ```

Typo.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Comment by alkemist** [comment]
Reviewed 2 of 13 files at r1, 1 of 1 files at r2.
Review status: 3 of 15 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Comment by johnrengelman** [comment]
Review status: 3 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[ratpack-core/src/main/java/ratpack/service/DependsOn.java, line 70 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3rndyLkQ9G_ZO9OD:-KIe-O1-excv5TR4q-uK:-1906204241) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/DependsOn.java#L70)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Why was this needed?
</details>
These specs failed when I had a ratpack app running when I ran them, they were trying to bind to 5050.

---

*[ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java, line 72 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3twZar7PpKmKebTo:-KIe-YfGEM-y3YQRftuT:1399668606) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java#L72)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Why was this needed?
</details>
same replace as above.

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4WoZu3-LZ4BRKDV5:-KIe-_hD0-78OBoio50Z:-1745225533) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-manual/src/content/chapters/38-retrofit.md#L50)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> I think we should be setting up these objects in the registry for these examples, as that's closer to reality and provides a better hint on what people should do.
</details>
I agree, but in this example I need to get the `PublicAddress` which is only resolvable during a request. Got a suggestion on how to make that work?

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java, line 96 \[r1\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KGsOorVnHsVhWkXxoye-r1-96:-KIXmCQzXXFBqSmH64JS:-387499785) ([raw file](https://github.com/ratpack/ratpack/blob/9ebc95889df11d68d7e38cddc0e00e8249c0013d/ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java#L96)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Do away with what? asking for a URI?
</details>
That was an old comment from before I refactored the API a bit.
Though I could see the usefullness of being able to provide a specific `HttpClient` if these changes happen (https://github.com/ratpack/ratpack/pull/979)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Comment by johnrengelman** [comment]
Review status: 3 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4WoZu3-LZ4BRKDV5:-KIeUengFONBEco5QyvA:716421725) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-manual/src/content/chapters/38-retrofit.md#L50)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> I agree, but in this example I need to get the `PublicAddress` which is only resolvable during a request. Got a suggestion on how to make that work?
</details>
I reworked the examples to use a 2nd `EmbeddedApp` and place the client proxy into the registry in the main ap.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Comment by alkemist** [comment]
Looks really good man, great job. 

I think we could benefit with just a little bit more of an explanation as to why you would use this instead of the plain HTTP client in the manual.

<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> @ratpack/committers would appreciate some feedback, especially around the API in `RatpackRetrofit`
</details>


---

Review status: 3 of 15 files reviewed at latest revision, 8 unresolved discussions.

---

*[ratpack-core/src/main/java/ratpack/service/DependsOn.java, line 70 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3rndyLkQ9G_ZO9OD:-KIgeOQ0FbiNlfSH6mew:160383661) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/DependsOn.java#L70)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> These specs failed when I had a ratpack app running when I ran them, they were trying to bind to 5050.
</details>
OK.

---

*[ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java, line 72 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX3twZar7PpKmKebTo:-KIgeVWZJUk4osuim2vr:160383661) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-core/src/main/java/ratpack/service/ServiceDependencies.java#L72)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> same replace as above.
</details>
OK.

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4WoZu3-LZ4BRKDV5:-KIgej2fukzzJpPZmRdB:305813840) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-manual/src/content/chapters/38-retrofit.md#L50)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> I reworked the examples to use a 2nd `EmbeddedApp` and place the client proxy into the registry in the main ap.
</details>
Nice solution. That API app won't be getting shutdown though.

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 3 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIghLCElghaWBlCgMBY:-KIghLCElghaWBlCgMBZ:-1877627590) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-manual/src/content/chapters/38-retrofit.md#L3)):*
> ```Markdown
> # retrofit
> 
> The `ratpack-retrofit2` extension provides integration for declarative HTTP clients with the [retrofit2](http://square.github.io/retrofit) library.
> ```

Would be good to include the Retrofit version the integration is based on here. We do this for some other stuff.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallAdapterFactory.java, line 32 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIggOCgc3bqfFNuIgmr:-KIggOCgc3bqfFNuIgms:54339629) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallAdapterFactory.java#L32)):*
> ```Java
> public class RatpackCallAdapterFactory extends CallAdapter.Factory {
> 
>   public static final RatpackCallAdapterFactory INSTANCE = new RatpackCallAdapterFactory();
> ```

Should have a private constructor if usage should only be via this singleton.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java, line 41 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIggIyP2Tqc6VYwhXjs:-KIggIyP2Tqc6VYwhXjt:-751890149) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java#L41)):*
> ```Java
>   public static final RatpackCallFactory INSTANCE = new RatpackCallFactory();
> 
>   public RatpackCallFactory() {
> ```

This can be private.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java, line 83 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIggYvsW9b_zLKpYj6i:-KIggYvsW9b_zLKpYj6j:755165412) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java#L83)):*
> ```Java
>     Promise<ReceivedResponse> promise() {
>       final Execution e = Execution.current();
>       //Get an HttpClient from the Execution if it exists, else get it from the Context
> ```

Should just read from the context, as it will try the execution first and then try the context registry and then the app registry.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/Utils.java, line 24 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIgh7bymuMQp6sDoQzP:-KIgh7bymuMQp6sDoQzQ:377791671) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/Utils.java#L24)):*
> ```Java
> import java.util.Arrays;
> 
> public class Utils {
> ```

Should be abstract.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976#-:-KIghB279_lLsO5q2Aej:724493245)*

**Comment by johnrengelman** [comment]
Review status: 3 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[ratpack-manual/src/content/chapters/38-retrofit.md, line 50 \[r3\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIX4WoZu3-LZ4BRKDV5:-KIh1wuoYpJJ9ZcnB1ad:-1531864576) ([raw file](https://github.com/ratpack/ratpack/blob/5498d95e4aa81e32ee4fdfe0fc99002d98e124e1/ratpack-manual/src/content/chapters/38-retrofit.md#L50)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Nice solution. That API app won't be getting shutdown though.
</details>
I added explicit `api.close()` calls at the end of .`test(Action`.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java, line 83 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIggYvsW9b_zLKpYj6i:-KIh0swSqSir7Tr47o-P:-1454191020) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java#L83)):*
<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Should just read from the context, as it will try the execution first and then try the context registry and then the app registry.
</details>
So just go `Execution.current().get(Context.class).get(HttpClient.class)`?

Is there a way to put a dummy `Context` then in the spec? `DefaultContext` seems to take a lot of setup.

---

*[ratpack-retrofit2/src/test/groovy/ratpack/retrofit/RatpackRetrofitSpec.groovy, line 75 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIh1D1tTH2xlDz9BFgc:-KIh1D1uzUGPNFJK2RyP:-1775284853) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/test/groovy/ratpack/retrofit/RatpackRetrofitSpec.groovy#L75)):*
> ```Groovy
>     def value = ExecHarness.yieldSingle({ RegistrySpec r ->
>       r.add(client)
>     }) {
> ```

This is what I was doing to test without spinning up a whole app.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Comment by johnrengelman** [comment]
I reworked the manual a bit trying to get the point across.

<details><summary><i>Previously, alkemist (Luke Daley) wrote…</i></summary>
> Looks really good man, great job. 
> 
> I think we could benefit with just a little bit more of an explanation as to why you would use this instead of the plain HTTP client in the manual.
</details>


---

Review status: 3 of 16 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976#-:-KIhHAtdvV1gWS5IGjcM:-1662354402)*

**Comment by johnrengelman** [comment]
Review status: 2 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java, line 83 \[r4\]](https://reviewable.io:443/reviews/ratpack/ratpack/976#-KIggYvsW9b_zLKpYj6i:-KIsWjPdADrGA81l-wuI:-1191059050) ([raw file](https://github.com/ratpack/ratpack/blob/e34ed5970b21ee7db44cc4e541f155292f8f0a62/ratpack-retrofit2/src/main/java/ratpack/retrofit/internal/RatpackCallFactory.java#L83)):*
<details><summary><i>Previously, johnrengelman (John Engelman) wrote…</i></summary>
> So just go `Execution.current().get(Context.class).get(HttpClient.class)`?
> 
> Is there a way to put a dummy `Context` then in the spec? `DefaultContext` seems to take a lot of setup.
</details>
I fixed this my using a `Mock(Context)` in the spec.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/ratpack/ratpack/976)*

**Pull request merged**